shared/health/src/main/java/io/pravega/shared/health/impl/CompositeHealthContributor.java

```diff
@@ -0,0 +1,121 @@
+ */
+package io.pravega.shared.health.impl;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.pravega.shared.health.ContributorRegistry;
+import io.pravega.shared.health.Health;
+import io.pravega.shared.health.HealthContributor;
+import io.pravega.shared.health.Status;
+import io.pravega.shared.health.StatusAggregator;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Objects;
+
+/**
+ * The {@link CompositeHealthContributor} class defines the base logic required to build {@link HealthContributor} objects
+ * that depend on (or are composed of) other {@link HealthContributor} objects. To do this we need to due two things:
+ * reducing/reconciling the (potentially) many different {@link Status} states and a way to persist these relationships.
+ *
+ * {@link CompositeHealthContributor} objects are used *strictly* for grouping, meaning that they do not define their own
+ * health checking logic and therefore does not export any details ({@link Health#getDetails}).
+ */
+@Slf4j
+@ThreadSafe
+public abstract class CompositeHealthContributor implements HealthContributor {
+
+    /**
+     * The {@link StatusAggregator} used to perform the aggregation of all the {@link HealthContributor} dependencies.
+     */
+    @Getter
+    private final StatusAggregator aggregator;
+
+    private final ContributorRegistry registry;
+
+    private final Collection<HealthContributor> contributors = ImmutableSet.of();
+
+    CompositeHealthContributor(StatusAggregator aggregator, ContributorRegistry registry) {
+        this.aggregator = aggregator;
+        this.registry = registry;
+    }
+
+    /**
+     * Requests the {@link Health} without fetching any details from it's dependencies.
+     * @return The {@link Health} result.
+     */
+    public Health getHealthSnapshot() {
+        return getHealthSnapshot(false);
+    }
+
+    /**
+     * Performs a {link Health} check on *all* descendant {@link HealthContributor}. The 'includeDetails' flag
+     * has different semantics depending on the type of sub-class.
+     *
+     * {@link io.pravega.shared.health.HealthIndicator} object will be requested to fetch their {@link Health#getDetails()}
+     * while {@link HealthComponent} will provide the {@link Health} results for all it's descendants.
+     *
+     * @param includeDetails Whether or not to request details from each of the {@link HealthContributor}.
+     * @return The {@link Health} result of the {@link CompositeHealthContributor}.
+     */
+    public Health getHealthSnapshot(boolean includeDetails) {
+        Health.HealthBuilder builder = Health.builder().name(getName());
+        Collection<Status> statuses = new ArrayList<>();
+        // Fetch the Health Status of all dependencies.
+        val children =  contributors().stream()
```

Comment:
Why do you generate `children` if in the last line of this method you may not include it if `includeDetails` is false?

Reply:
This iteration over all children must still be done to build the aggregate `Status`, so I figured  it would not be that wasteful to simply return the children `Health` results in case the details are requested.

Is something like this better?:
```
        Collection<Status> statuses = new ArrayList<>();
        List<Health> children = new ArrayList<>();
        getContributors().stream()
                .filter(Objects::nonNull)
                .forEach(contributor -> {
                    Health health = contributor.getHealthSnapshot(includeDetails);
                    statuses.add(health.getStatus());
                    if (includeDetails) {
                        children.add(health);
                    }
                });
```